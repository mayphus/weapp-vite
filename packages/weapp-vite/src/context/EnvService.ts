import type { ConfigService } from './ConfigService'
import { inject, injectable } from 'inversify'
import { Symbols } from './Symbols'

@injectable()
export class EnvService {
  /**
   * esbuild 定义的环境变量
   */
  defineEnv: Record<string, any>
  constructor(
    @inject(Symbols.ConfigService)
    private readonly configService: ConfigService,
  ) {
    this.defineEnv = {} // 初始化定义的环境变量对象
  }

  get defineImportMetaEnv() {
    const env = {
      MP_PLATFORM: this.configService.options?.platform,
      ...this.defineEnv,
    }
    const define: Record<string, any> = {}
    for (const [key, value] of Object.entries(env)) {
      define[`import.meta.env.${key}`] = JSON.stringify(value)
    }

    define[`import.meta.env`] = JSON.stringify(env)
    return define
  }

  setDefineEnv(key: string, value: any) {
    this.defineEnv[key] = value
  }
}
