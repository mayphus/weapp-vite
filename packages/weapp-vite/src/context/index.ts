export * from './CompilerContext'
export * from './services'
