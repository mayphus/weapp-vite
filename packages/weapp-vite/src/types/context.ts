export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
  attrs: Record<string, string>
}

export interface ScanComponentItem {
  start: number
  end: number
}

export type ComponentsMap = Record<string, ScanComponentItem[]>
