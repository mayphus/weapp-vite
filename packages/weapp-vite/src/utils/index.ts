export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
