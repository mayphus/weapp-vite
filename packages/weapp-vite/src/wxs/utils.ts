import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
