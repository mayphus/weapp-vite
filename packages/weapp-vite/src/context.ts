import type { PackageJson } from 'pkg-types'
import type { RollupOutput, RollupWatcher } from 'rollup'
import type { AppEntry, CompilerContextOptions, Entry, EntryJsonFragment, ProjectConfig, ResolvedAlias, SubPackage, SubPackageMetaValue, TsupOptions } from './types'
import { createRequire } from 'node:module'
import process from 'node:process'
import { addExtension, defu, get, isObject, objectHash, removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import path from 'pathe'
import { build, type InlineConfig, loadConfigFromFile } from 'vite'
import tsconfigPaths from 'vite-tsconfig-paths'
import { defaultExcluded } from './defaults'
import logger from './logger'
import { vitePluginWeapp } from './plugins'
import { changeFileExtension, findJsEntry, getAliasEntries, getProjectConfig, readCommentJson, resolveImportee } from './utils'
import './config'

const require = createRequire(import.meta.url)

let logBuildAppFinishOnlyShowOnce = false

function logBuildAppFinish() {
  if (!logBuildAppFinishOnlyShowOnce) {
    logger.success('应用构建完成！预览方式 ( `2` 种选其一即可)：')
    logger.info('执行 `npm run open` / `yarn open` / `pnpm open` 直接在 `微信开发者工具` 里打开当前应用')
    logger.info('或手动打开微信开发者工具,导入根目录(`project.config.json` 文件所在的目录),即可预览效果')
    logBuildAppFinishOnlyShowOnce = true
  }
}

function logBuildIndependentSubPackageFinish(root: string) {
  logger.success(`独立分包 ${root} 构建完成！`)
}
export class CompilerContext {
  /**
   * loadDefaultConfig 的时候会被重新赋予
   */
  inlineConfig: InlineConfig
  cwd: string
  isDev: boolean
  projectConfig: ProjectConfig
  mode: string
  packageJson: PackageJson

  private rollupWatcherMap: Map<string, RollupWatcher>

  entriesSet: Set<string>
  entries: Entry[]

  appEntry?: AppEntry

  subPackageMeta: Record<string, SubPackageMetaValue>

  aliasEntries: ResolvedAlias[]

  constructor(options?: CompilerContextOptions) {
    const { cwd, isDev, inlineConfig, projectConfig, mode, packageJson } = defu<Required<CompilerContextOptions>, CompilerContextOptions[]>(options, {
      cwd: process.cwd(),
      isDev: false,
      projectConfig: {},
      inlineConfig: {},
      packageJson: {},
    })
    this.cwd = cwd
    this.inlineConfig = inlineConfig
    this.isDev = isDev
    this.projectConfig = projectConfig
    this.mode = mode
    this.packageJson = packageJson
    this.rollupWatcherMap = new Map()
    this.subPackageMeta = {}
    this.entriesSet = new Set()
    this.entries = []
    this.aliasEntries = []
  }

  get srcRoot() {
    return this.inlineConfig?.weapp?.srcRoot ?? ''
  }

  relativeSrcRoot(p: string) {
    if (this.srcRoot) {
      return path.relative(this.srcRoot, p)
    }
    return p
  }

  /**
   * @description 写在 projectConfig 里面的 miniprogramRoot / srcMiniprogramRoot
   * 默认为 'dist'
   *
   */
  get mpDistRoot(): string | undefined {
    return this.projectConfig.miniprogramRoot || this.projectConfig.srcMiniprogramRoot
  }

  private async internalDev(inlineConfig: InlineConfig) {
    const watcher = (
      await build(
        inlineConfig,
      )
    ) as RollupWatcher
    await new Promise((resolve, reject) => {
      watcher.on('event', async (e) => {
        if (e.code === 'END') {
          await this.buildSubPackage()
          logBuildAppFinish()
          resolve(e)
        }
        else if (e.code === 'ERROR') {
          reject(e)
        }
      })
    })
    this.setRollupWatcher(watcher)
    return watcher
  }

  async runDev() {
    if (process.env.NODE_ENV === undefined) {
      process.env.NODE_ENV = 'development'
    }

    const inlineConfig = this.getConfig()

    const watcher = await this.internalDev(inlineConfig)

    return watcher
  }

  getConfig(subPackageMeta?: SubPackageMetaValue, ...configs: Partial<InlineConfig>[]) {
    if (this.isDev) {
      return defu<InlineConfig, InlineConfig[]>(
        this.inlineConfig,
        ...configs,
        {
          root: this.cwd,
          mode: 'development',
          plugins: [vitePluginWeapp(this, subPackageMeta)],
          build: {
            watch: {
              exclude: [
                ...defaultExcluded,
                this.mpDistRoot ? path.join(this.mpDistRoot, '**') : 'dist/**',
              ],
              chokidar: {
                ignored: [...defaultExcluded],
              },
            },
            minify: false,
            emptyOutDir: false,
          },
        },
      )
    }
    else {
      const inlineConfig = defu<InlineConfig, InlineConfig[]>(
        this.inlineConfig,
        ...configs,
        {
          root: this.cwd,
          plugins: [vitePluginWeapp(this, subPackageMeta)],
          mode: 'production',
          build: {
            emptyOutDir: false,
          },
        },
      )
      inlineConfig.logLevel = 'info'
      return inlineConfig
    }
  }

  async runProd() {
    if (this.mpDistRoot) {
      await fs.emptyDir(path.resolve(this.cwd, this.mpDistRoot))
      logger.success(`已清空 ${this.mpDistRoot} 目录`)
    }
    const output = (await build(
      this.getConfig(),
    ))
    await this.buildSubPackage()
    logBuildAppFinish()
    return output as RollupOutput | RollupOutput[]
  }

  async build() {
    if (this.isDev) {
      await this.runDev()
    }
    else {
      await this.runProd()
    }
  }

  async loadDefaultConfig() {
    const projectConfig = await getProjectConfig(this.cwd)
    this.projectConfig = projectConfig
    if (!this.mpDistRoot) {
      logger.error('请在 `project.config.json` 里设置 `miniprogramRoot`, 比如可以设置为 `dist/` ')
      return
    }
    const packageJsonPath = path.resolve(this.cwd, 'package.json')
    const external: string[] = []
    if (await fs.exists(packageJsonPath)) {
      const localPackageJson: PackageJson = await fs.readJson(packageJsonPath, {
        throws: false,
      }) || {}
      this.packageJson = localPackageJson
      if (localPackageJson.dependencies) {
        external.push(...Object.keys(localPackageJson.dependencies))
      }
    }

    const loaded = await loadConfigFromFile({
      command: this.isDev ? 'serve' : 'build',
      mode: this.mode,
    }, undefined, this.cwd)

    this.inlineConfig = defu<InlineConfig, (InlineConfig | undefined)[]>({
      configFile: false,
    }, loaded?.config, {
      build: {
        rollupOptions: {
          output: {
            format: 'cjs',
            strict: false,
            entryFileNames: (chunkInfo) => {
              const name = this.relativeSrcRoot(chunkInfo.name)
              if (name.endsWith('.ts')) {
                const baseFileName = removeExtension(name)
                if (baseFileName.endsWith('.wxs')) {
                  return baseFileName
                }
                return addExtension(baseFileName, '.js')
              }
              return name
            },
          },
          external,
        },
        assetsDir: '.',
        commonjsOptions: {
          transformMixedEsModules: true,
          include: undefined,
        },
      },
      plugins: [
        tsconfigPaths(),
      ],
      logLevel: 'warn',
    })
    // this.inlineConfig.plugins?.push(
    //   viteStaticCopy({
    //     targets: [
    //       {
    //         src: path.join(this.inlineConfig.weapp?.srcRoot ?? '', this.projectConfig.sitemapLocation),
    //         dest: '.',
    //       },
    //     ],
    //   }),
    // )
    this.aliasEntries = getAliasEntries(this.inlineConfig.weapp?.jsonAlias)
  }

  get dependenciesCacheFilePath() {
    return path.resolve(this.cwd, 'node_modules/weapp-vite/.cache/npm.json')
  }

  get dependenciesCacheHash() {
    return objectHash(this.packageJson.dependencies ?? {})
  }

  writeDependenciesCache() {
    return fs.outputJSON(this.dependenciesCacheFilePath, {
      '/': this.dependenciesCacheHash,
    })
  }

  async readDependenciesCache() {
    if (await fs.exists(this.dependenciesCacheFilePath)) {
      return await fs.readJson(this.dependenciesCacheFilePath, { throws: false })
    }
  }

  async checkDependenciesCacheOutdate() {
    const json = await this.readDependenciesCache()
    if (isObject(json)) {
      return this.dependenciesCacheHash !== json['/']
    }
    return true
  }

  // https://cn.vitejs.dev/guide/build.html#library-mode
  // miniprogram_dist
  // miniprogram
  // https://developers.weixin.qq.com/miniprogram/dev/devtools/npm.html#%E8%87%AA%E5%AE%9A%E4%B9%89%E7%BB%84%E4%BB%B6%E7%9B%B8%E5%85%B3%E7%A4%BA%E4%BE%8B
  async buildNpm(options?: TsupOptions) {
    const { build: tsupBuild } = await import('tsup')
    const isDependenciesCacheOutdate = await this.checkDependenciesCacheOutdate()
    // if (!await this.isDependenciesCacheOutdate()) {
    //   logger.success(`依赖未发生变化，跳过处理!`)

    // }

    let packNpmRelationList: {
      packageJsonPath: string
      miniprogramNpmDistDir: string
    }[] = []
    if (this.projectConfig.setting?.packNpmManually && Array.isArray(this.projectConfig.setting.packNpmRelationList)) {
      packNpmRelationList = this.projectConfig.setting.packNpmRelationList
    }
    else {
      packNpmRelationList = [
        {
          miniprogramNpmDistDir: '.',
          packageJsonPath: './package.json',
        },
      ]
    }
    for (const relation of packNpmRelationList) {
      const packageJsonPath = path.resolve(this.cwd, relation.packageJsonPath)
      if (await fs.exists(packageJsonPath)) {
        const pkgJson: PackageJson = await fs.readJson(packageJsonPath)
        const outDir = path.resolve(this.cwd, relation.miniprogramNpmDistDir, 'miniprogram_npm')
        if (pkgJson.dependencies) {
          const dependencies = Object.keys(pkgJson.dependencies)
          if (dependencies.length > 0) {
            for (const dep of dependencies) {
              const id = `${dep}/package.json`
              const targetJson = require(id)

              if (Reflect.has(targetJson, 'miniprogram') && targetJson.miniprogram) {
                const targetJsonPath = require.resolve(id)
                const dest = path.join(outDir, dep)
                if (!isDependenciesCacheOutdate && await fs.exists(dest)) {
                  logger.success(`${dep} 依赖未发生变化，跳过处理!`)
                  continue
                }
                await fs.copy(
                  path.resolve(
                    path.dirname(targetJsonPath),
                    targetJson.miniprogram,
                  ),
                  dest,
                )
              }
              else {
                const destOutDir = path.join(outDir, dep)
                if (!isDependenciesCacheOutdate && await fs.exists(destOutDir)) {
                  logger.success(`${dep} 依赖未发生变化，跳过处理!`)
                  continue
                }

                const mergedOptions: TsupOptions = defu<TsupOptions, TsupOptions[]>(options, {
                  entry: {
                    index: require.resolve(dep),
                  },
                  format: ['cjs'],
                  outDir: destOutDir,
                  silent: true,
                  shims: true,
                  outExtension: () => {
                    return {
                      js: '.js',
                    }
                  },
                  sourcemap: false,
                  // clean: false,
                })
                const resolvedOptions = this.inlineConfig.weapp?.npm?.tsup?.(mergedOptions)
                let finalOptions: TsupOptions | undefined
                if (resolvedOptions === undefined) {
                  finalOptions = mergedOptions
                }
                else if (isObject(resolvedOptions)) {
                  finalOptions = resolvedOptions
                }
                finalOptions && await tsupBuild(finalOptions)
              }
              logger.success(`${dep} 依赖处理完成!`)
            }
          }
        }
      }
    }
    await this.writeDependenciesCache()
  }

  private async usingComponentsHandler(entry: EntryJsonFragment, relDir: string) {
    // this.packageJson.dependencies
    const { usingComponents } = entry.json as unknown as {
      usingComponents: Record<string, string>
    }
    if (usingComponents) {
      for (const [, componentUrl] of Object.entries(usingComponents)) {
        if (/plugin:\/\//.test(componentUrl)) {
          // console.log(`发现插件 ${usingComponent}`)
          continue
        }
        const tokens = componentUrl.split('/')
        // 来自 dependencies 的依赖直接跳过
        if (tokens[0] && isObject(this.packageJson.dependencies) && Reflect.has(this.packageJson.dependencies, tokens[0])) {
          continue
        }
        // start with '/' 表述默认全局别名
        else if (tokens[0] === '') {
          await this.scanComponentEntry(componentUrl.substring(1), path.resolve(this.cwd, this.srcRoot))
        }
        else {
          const importee = resolveImportee(componentUrl, entry, this.aliasEntries)
          await this.scanComponentEntry(importee, relDir)
        }
      }
    }
  }

  resetEntries() {
    this.entriesSet.clear()
    this.entries.length = 0
    this.subPackageMeta = {}
  }

  async scanAppEntry() {
    this.resetEntries()
    const appDirname = path.resolve(this.cwd, this.srcRoot)
    const appConfigFile = path.resolve(appDirname, 'app.json')
    const appEntryPath = await findJsEntry(appConfigFile)
    // https://developers.weixin.qq.com/miniprogram/dev/framework/structure.html
    // js + json
    if (appEntryPath && await fs.exists(appConfigFile)) {
      const config = await readCommentJson(appConfigFile) as unknown as {
        pages: string[]
        usingComponents: Record<string, string>
        subpackages: SubPackage[]
        subPackages: SubPackage[]
        themeLocation?: string
        sitemapLocation?: string
      }
      if (isObject(config)) {
        if (this.entriesSet.has(appEntryPath)) {
          return
        }
        this.entriesSet.add(appEntryPath)
        const appEntry: AppEntry = {
          path: appEntryPath,
          json: config,
          jsonPath: appConfigFile,
          type: 'app',
        }
        this.entries.push(appEntry)
        this.appEntry = appEntry

        const { pages, subpackages = [], subPackages = [], sitemapLocation = 'sitemap.json', themeLocation = 'theme.json' } = config
        // sitemap.json
        if (sitemapLocation) {
          const sitemapJsonPath = path.resolve(appDirname, sitemapLocation)
          if (await fs.exists(sitemapJsonPath)) {
            appEntry.sitemapJsonPath = sitemapJsonPath
            appEntry.sitemapJson = await readCommentJson(sitemapJsonPath) as unknown as object
          }
        }
        // theme.json
        if (themeLocation) {
          const themeJsonPath = path.resolve(appDirname, themeLocation)
          if (await fs.exists(themeJsonPath)) {
            appEntry.themeJsonPath = themeJsonPath
            appEntry.themeJson = await readCommentJson(themeJsonPath) as unknown as object
          }
        }
        // https://developers.weixin.qq.com/miniprogram/dev/framework/subpackages/basic.html
        // 优先 subPackages
        const subs: SubPackage[] = [...subpackages, ...subPackages]
        // 组件
        await this.usingComponentsHandler(appEntry, appDirname)
        // 页面
        if (Array.isArray(pages)) {
          for (const page of pages) {
            await this.scanComponentEntry(page, appDirname)
          }
        }
        // 分包
        for (const sub of subs) {
          // 独立分包
          if (sub.independent) {
            const meta: SubPackageMetaValue = {
              entries: [],
              entriesSet: new Set(),
              subPackage: sub,
            }
            const scanComponentEntry = this.scanComponentEntry.bind(meta)
            if (Array.isArray(sub.pages)) {
              for (const page of sub.pages) {
                await scanComponentEntry(path.join(sub.root, page), appDirname)
              }
            }
            if (sub.entry) {
              await scanComponentEntry(path.join(sub.root, sub.entry), appDirname)
            }
            this.subPackageMeta[sub.root] = meta
          }
          else {
            // 普通分包
            if (Array.isArray(sub.pages)) {
              for (const page of sub.pages) {
                await this.scanComponentEntry(path.join(sub.root, page), appDirname)
              }
            }
            if (sub.entry) {
              await this.scanComponentEntry(path.join(sub.root, sub.entry), appDirname)
            }
          }
        }
        // 自定义 tabBar
        // https://developers.weixin.qq.com/miniprogram/dev/framework/ability/custom-tabbar.html
        if (get(appEntry, 'json.tabBar.custom')) {
          await this.scanComponentEntry('custom-tab-bar/index', appDirname)
        }
        return appEntry
      }
    }
    else {
      throw new Error(`在 ${appDirname} 目录下没有找到 \`app.json\`, 请确保你初始化了小程序项目，或者在 \`vite.config.ts\` 中设置的正确的 \`weapp.srcRoot\` 配置路径  `)
    }
  }

  // usingComponents
  // subpackages / subPackages
  // pages
  // https://developers.weixin.qq.com/miniprogram/dev/framework/structure.html
  // 页面可以没有 JSON
  async scanComponentEntry(componentEntry: string, dirname: string) {
    const entry = path.resolve(dirname, componentEntry)
    const jsEntry = await findJsEntry(entry)
    const partialEntry: Entry = {
      path: jsEntry!,
    }
    if (jsEntry && !this.entriesSet.has(jsEntry)) {
      this.entriesSet.add(jsEntry)
      this.entries.push(partialEntry)
    }
    const configFile = changeFileExtension(entry, 'json')
    if (await fs.exists(configFile)) {
      const config = await readCommentJson(configFile) as unknown as {
        usingComponents: Record<string, string>
      }
      const jsonFragment = {
        json: config,
        jsonPath: configFile,
      }
      if (jsEntry) {
        partialEntry.json = jsonFragment.json
        partialEntry.jsonPath = jsonFragment.jsonPath
      }
      if (isObject(config)) {
        await this.usingComponentsHandler(jsonFragment, path.dirname(configFile))
      }
    }
  }

  setRollupWatcher(watcher: RollupWatcher, root: string = '/') {
    const oldWatcher = this.rollupWatcherMap.get(root)
    oldWatcher?.close()
    this.rollupWatcherMap.set(root, watcher)
  }

  // 独立分包需要单独打包
  async buildSubPackage() {
    for (const [root, meta] of Object.entries(this.subPackageMeta)) {
      const inlineConfig = this.getConfig(meta, {
        build: {
          rollupOptions: {
            output: {
              chunkFileNames() {
                return `${root}/[name]-[hash].js`
              },
            },
          },
        },
      })
      const output = (await build(
        inlineConfig,
      ))
      if (this.isDev) {
        const watcher = output as RollupWatcher
        this.setRollupWatcher(watcher, root)
        const e = await new Promise((resolve, reject) => {
          watcher.on('event', (e) => {
            if (e.code === 'END') {
              logBuildIndependentSubPackageFinish(root)
              resolve(e)
            }
            else if (e.code === 'ERROR') {
              reject(e)
            }
          })
        })
        return e
      }
      else {
        logBuildIndependentSubPackageFinish(root)
      }
    }
  }
}

export async function createCompilerContext(options?: CompilerContextOptions) {
  const ctx = new CompilerContext(options)
  await ctx.loadDefaultConfig()
  return ctx
}
