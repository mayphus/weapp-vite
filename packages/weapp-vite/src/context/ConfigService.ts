import type { PackageJson } from 'pkg-types'
import type { OutputExtensions } from '../defaults'
import process from 'node:process'
import { addExtension, defu, removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import { injectable } from 'inversify'
import path from 'pathe'
import { type InlineConfig, loadConfigFromFile } from 'vite'
import tsconfigPaths from 'vite-tsconfig-paths'
import { getOutputExtensions, getWeappViteConfig } from '../defaults'
import { getAliasEntries, getProjectConfig } from '../utils'
import { logger } from './shared'

export interface LoadConfigOptions {
  cwd: string
  isDev: boolean
  mode: string
}

export type LoadConfigResult = NonNullable<Awaited<ReturnType<typeof loadConfig>>>

export async function loadConfig(opts: LoadConfigOptions) {
  const { cwd, isDev, mode } = opts
  const projectConfig = await getProjectConfig(cwd)
  const mpDistRoot = projectConfig.miniprogramRoot ?? projectConfig.srcMiniprogramRoot
  if (!mpDistRoot) {
    logger.error('请在 `project.config.json` 里设置 `miniprogramRoot`, 比如可以设置为 `dist/` ')
    return
  }
  const packageJsonPath = path.resolve(cwd, 'package.json')
  const external: (string | RegExp)[] = []
  let packageJson: PackageJson = {}
  if (await fs.exists(packageJsonPath)) {
    const localPackageJson: PackageJson = await fs.readJson(packageJsonPath, {
      throws: false,
    }) || {}
    packageJson = localPackageJson
    if (localPackageJson.dependencies) {
      external.push(...Object.keys(localPackageJson.dependencies))
    }
  }

  const loaded = await loadConfigFromFile({
    command: isDev ? 'serve' : 'build',
    mode,
  }, undefined, cwd)

  const loadedConfig = loaded?.config

  const srcRoot = loadedConfig?.weapp?.srcRoot ?? ''
  function relativeSrcRoot(p: string) {
    if (srcRoot) {
      return path.relative(srcRoot, p)
    }
    return p
  }

  const config = defu<InlineConfig, (InlineConfig | undefined)[]>({
    configFile: false,
  }, loadedConfig, {
    build: {
      rollupOptions: {
        output: {
          format: 'cjs',
          strict: false,
          entryFileNames: (chunkInfo) => {
            const name = relativeSrcRoot(chunkInfo.name)
            if (name.endsWith('.ts')) {
              const baseFileName = removeExtension(name)
              if (baseFileName.endsWith('.wxs')) {
                return baseFileName
              }
              return addExtension(baseFileName, '.js')
            }
            return name
          },
        },
        external,
      },
      assetsDir: '.',
      commonjsOptions: {
        transformMixedEsModules: true,
        include: undefined,
      },
    },
    logLevel: 'warn',
    weapp: getWeappViteConfig(),
  })

  const platform = config.weapp?.platform ?? 'weapp'

  const outputExtensions = getOutputExtensions(platform)
  config.plugins ??= []
  config.plugins?.push(tsconfigPaths(config.weapp?.tsconfigPaths))
  const aliasEntries = getAliasEntries(config.weapp?.jsonAlias)

  return {
    config,
    aliasEntries,
    outputExtensions,
    packageJson,
    relativeSrcRoot,
    cwd,
    isDev,
    mode,
    projectConfig,
    mpDistRoot,
    packageJsonPath,
    platform,
  }
}

@injectable()
export class ConfigService {
  options?: LoadConfigResult

  outputExtensions?: OutputExtensions
  constructor() {
    this.options = undefined
    this.outputExtensions = undefined
  }

  async load(options?: Partial<LoadConfigOptions>) {
    const opts = defu<LoadConfigOptions, LoadConfigOptions[]>(options, {
      cwd: process.cwd(),
      isDev: false,
      mode: 'development',
    })
    const opt = await loadConfig(opts)

    const _opts = defu<Required<LoadConfigResult>, Partial<LoadConfigResult>[]>(opt, {
      cwd: process.cwd(), // 当前工作目录，默认为进程的当前目录
      isDev: false, // 是否为开发模式，默认为false
      projectConfig: {}, // 项目配置对象，默认为空对象
      config: {}, // 内联配置对象，默认为空对象
      packageJson: {}, // package.json内容对象，默认为空对象
      platform: 'weapp', // 目标平台，默认为微信小程序平台
    })
    this.options = _opts
    this.outputExtensions = getOutputExtensions(_opts.platform) // 根据平台获取输出文件扩展名
    return _opts
  }
}
