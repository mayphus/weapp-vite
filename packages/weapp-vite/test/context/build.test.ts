import { createCompilerContext } from '@/context'
// import { omit } from 'lodash'

import { getFixture } from '../utils'

describe('build', () => {
  it('compilerContext', async () => {
    const ctx = await createCompilerContext({
      cwd: getFixture('mixjs'),
    })
    // expect(omit(ctx, 'cwd')).matchSnapshot()

    // expect(omit(ctx, 'cwd')).matchSnapshot()
    await ctx.runProd()
  })
})
