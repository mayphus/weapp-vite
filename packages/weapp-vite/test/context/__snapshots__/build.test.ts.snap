// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build basic > dist > app.js 1`] = `
"App({globalData:{},onLaunch(){}});
"
`;

exports[`build basic > dist > app.json 1`] = `
"{
  "pages": [
    "pages/index/index",
    "pages/index/vue"
  ],
  "window": {},
  "style": "v2",
  "componentFramework": "glass-easel",
  "sitemapLocation": "sitemap.json"
}"
`;

exports[`build basic > dist > app.wxss 1`] = `
".app{color:red}
/*$vite$:1*/"
`;

exports[`build basic > dist > components/Navbar/Navbar.js 1`] = `
"Component({});
"
`;

exports[`build basic > dist > components/Navbar/Navbar.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`build basic > dist > components/Navbar/Navbar.wxml 1`] = `
"<view>Navbar</view>
"
`;

exports[`build basic > dist > components/Navbar/Navbar.wxss 1`] = `
".navbar{color:red}
/*$vite$:1*/"
`;

exports[`build basic > dist > pages/index/index.js 1`] = `
"Page({});
"
`;

exports[`build basic > dist > pages/index/index.json 1`] = `
"{
  "usingComponents": {
    "Navbar": "/components/Navbar/Navbar"
  },
  "navigationBarTitleText": "初始模板"
}"
`;

exports[`build basic > dist > pages/index/index.wxml 1`] = `
"<view>
  <Navbar />
  <view>Hello World From Weapp-vite!</view>
</view>
"
`;

exports[`build basic > dist > pages/index/index.wxss 1`] = `
".page{color:red}
/*$vite$:1*/"
`;

exports[`build basic > dist > sitemap.json 1`] = `
"{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [
    {
      "action": "allow",
      "page": "*"
    }
  ]
}"
`;

exports[`build basic > dist > theme.json 1`] = `
"{
  "light": {},
  "dark": {}
}"
`;

exports[`build basic > dist 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Navbar/Navbar.wxss",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "pages/index/index.wxss",
  "sitemap.json",
  "theme.json",
]
`;

exports[`tabbar-appbar > dist > app.js 1`] = `
"App({globalData:{},onLaunch(){}});
"
`;

exports[`tabbar-appbar > dist > app.json 1`] = `
"{
  "pages": [
    "pages/index/index",
    "pages/index/vue"
  ],
  "window": {},
  "style": "v2",
  "componentFramework": "glass-easel",
  "sitemapLocation": "sitemap.json",
  "tabBar": {
    "custom": true
  },
  "appBar": {},
  "usingComponents": {
    "Test": "/components/Test"
  }
}"
`;

exports[`tabbar-appbar > dist > app-bar/index.js 1`] = `
"Component({});
"
`;

exports[`tabbar-appbar > dist > app-bar/index.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`tabbar-appbar > dist > app-bar/index.wxml 1`] = `"<view>hello weapp-vite! from app-bar</view>"`;

exports[`tabbar-appbar > dist > app-bar/index.wxss 1`] = `
".ss{color:red}
/*$vite$:1*/"
`;

exports[`tabbar-appbar > dist > async2-D81jY5iE.js 1`] = `
"Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const n="async2",a=2;exports.async2=n;exports.default=a;
"
`;

exports[`tabbar-appbar > dist > components/Navbar/Navbar.js 1`] = `
"Component({});
"
`;

exports[`tabbar-appbar > dist > components/Navbar/Navbar.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`tabbar-appbar > dist > components/Navbar/Navbar.wxml 1`] = `
"<view>Navbar</view>
"
`;

exports[`tabbar-appbar > dist > components/Test/index.js 1`] = `
"Component({});
"
`;

exports[`tabbar-appbar > dist > components/Test/index.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`tabbar-appbar > dist > components/Test/index.wxml 1`] = `"<view>Navbar</view>"`;

exports[`tabbar-appbar > dist > components/Test/index.wxss 1`] = `
".test{clear:both}
/*$vite$:1*/"
`;

exports[`tabbar-appbar > dist > custom-tab-bar/index.js 1`] = `
"const _=require("../other-BcprOjgc.js"),e=_.__CJS__export_default__||_.__CJS__import__0__;Component({lifetimes:{attached(){console.log(e)}}});
"
`;

exports[`tabbar-appbar > dist > custom-tab-bar/index.json 1`] = `
"{
  "component": true,
  "styleIsolation": "apply-shared",
  "usingComponents": {}
}"
`;

exports[`tabbar-appbar > dist > custom-tab-bar/index.wxml 1`] = `"<view class="pb-8 pt-4 bg-[#af6ec5] flex justify-center">hello weapp-vite! aa</view>"`;

exports[`tabbar-appbar > dist > custom-tab-bar/index.wxss 1`] = `
".a{color:red}
/*$vite$:1*/"
`;

exports[`tabbar-appbar > dist > other-BcprOjgc.js 1`] = `
"var _={exports:{}};const t="other";_.exports={other:t};const e=(_.exports==null?{}:_.exports).default||_.exports,o=Object.freeze(Object.defineProperty({__proto__:null,default:e},Symbol.toStringTag,{value:"Module"}));exports.__CJS__export_default__=e;exports.__CJS__import__0__=o;
"
`;

exports[`tabbar-appbar > dist > pages/index/index.js 1`] = `
"const o=require("../../other-BcprOjgc.js"),_=3,{part:r}=_,{other:a}=o.__CJS__export_default__||o.__CJS__import__0__;Page({async onLoad(){const e=await require.async("./async"),t=await Promise.resolve().then(()=>require("../../async2-D81jY5iE.js"));console.log(r,a,e,t)}});
"
`;

exports[`tabbar-appbar > dist > pages/index/index.json 1`] = `
"{
  "usingComponents": {
    "Navbar": "/components/Navbar/Navbar"
  },
  "navigationBarTitleText": "初始模板"
}"
`;

exports[`tabbar-appbar > dist > pages/index/index.wxml 1`] = `
"<view>
  <Navbar />
  <view>Hello World From Weapp-vite!</view>
</view>
"
`;

exports[`tabbar-appbar > dist > sitemap.json 1`] = `
"{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [
    {
      "action": "allow",
      "page": "*"
    }
  ]
}"
`;

exports[`tabbar-appbar > dist > theme.json 1`] = `
"{
  "light": {},
  "dark": {}
}"
`;

exports[`tabbar-appbar > dist 1`] = `
[
  "app-bar/index.js",
  "app-bar/index.json",
  "app-bar/index.wxml",
  "app-bar/index.wxss",
  "app.js",
  "app.json",
  "async2-D81jY5iE.js",
  "components/Navbar/Navbar.js",
  "components/Navbar/Navbar.json",
  "components/Navbar/Navbar.wxml",
  "components/Test/index.js",
  "components/Test/index.json",
  "components/Test/index.wxml",
  "components/Test/index.wxss",
  "custom-tab-bar/index.js",
  "custom-tab-bar/index.json",
  "custom-tab-bar/index.wxml",
  "custom-tab-bar/index.wxss",
  "other-BcprOjgc.js",
  "pages/index/index.js",
  "pages/index/index.json",
  "pages/index/index.wxml",
  "sitemap.json",
  "theme.json",
]
`;
